Devirtualize calls to protocol composition type methods

zoecarver opened the pull request:
<!-- What's in this pull request? -->
This patch devirtualizes calls to methods of protocol composition types when those types are composed of a class and method. (when possible) It removes `init_existential_ref` and `open_existential_ref` instructions (existentials/existential unwrapping). This also allows for optimization of the methods directly (inlining, etc.).

Here's an example:
**Before:**
```sil
// witnessEntryPoint(c:)
sil @$s3run17witnessEntryPoint1cSiAA6ClassBC_tF : $@convention(thin) (@guaranteed ClassB) -> Int {
// %0                                             // users: %2, %1
bb0(%0 : $ClassB):
  debug_value %0 : $ClassB, let, name "c", argno 1 // id: %1
  %2 = init_existential_ref %0 : $ClassB : $ClassB, $ClassA<String> & ProtocolA // users: %4, %3
  debug_value %2 : $ClassA<String> & ProtocolA, let, name "x", argno 1 // id: %3
  %4 = open_existential_ref %2 : $ClassA<String> & ProtocolA to $@opened("154CD8E6-FE7C-11E9-9975-ACBC329C418B") ClassA<String> & ProtocolA // users: %8, %7, %6, %5
  %5 = alloc_stack $@opened("154CD8E6-FE7C-11E9-9975-ACBC329C418B") ClassA<String> & ProtocolA // type-defs: %4; users: %9, %8, %6
  store %4 to %5 : $*@opened("154CD8E6-FE7C-11E9-9975-ACBC329C418B") ClassA<String> & ProtocolA // id: %6
  %7 = witness_method $@opened("154CD8E6-FE7C-11E9-9975-ACBC329C418B") ClassA<String> & ProtocolA, #ProtocolA.foo!1 : <Self where Self : ProtocolA> (Self) -> () -> Int, %4 : $@opened("154CD8E6-FE7C-11E9-9975-ACBC329C418B") ClassA<String> & ProtocolA : $@convention(witness_method: ProtocolA) <τ_0_0 where τ_0_0 : ProtocolA> (@in_guaranteed τ_0_0) -> Int // type-defs: %4; user: %8
  %8 = apply %7<@opened("154CD8E6-FE7C-11E9-9975-ACBC329C418B") ClassA<String> & ProtocolA>(%5) : $@convention(witness_method: ProtocolA) <τ_0_0 where τ_0_0 : ProtocolA> (@in_guaranteed τ_0_0) -> Int // type-defs: %4; user: %10
  dealloc_stack %5 : $*@opened("154CD8E6-FE7C-11E9-9975-ACBC329C418B") ClassA<String> & ProtocolA // id: %9
  return %8 : $Int                                // id: %10
} // end sil function '$s3run17witnessEntryPoint1cSiAA6ClassBC_tF'
```

**After:**
```sil
// witnessEntryPoint(c:)
sil @$s3run17witnessEntryPoint1cSiAA6ClassBC_tF : $@convention(thin) (@guaranteed ClassB) -> Int {
bb0(%0 : $ClassB):
  %1 = integer_literal $Builtin.Int64, 10         // user: %2
  %2 = struct $Int (%1 : $Builtin.Int64)          // user: %3
  return %2 : $Int                                // id: %3
} // end sil function '$s3run17witnessEntryPoint1cSiAA6ClassBC_tF'
```

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Partially resolves [SR-11695](https://bugs.swift.org/browse/SR-11695) (doesn't optimize the overridden class -- I can work on that in another PR).

Once reviewed, can someone with commit access run swift-ci to test and benchmark, please?

zoecarver committed:
Implement devirtualization calls to composition type protocols

zoecarver committed:
Clean up implementation and remove debug

zoecarver committed:
Add a test

zoecarver committed:
Cleanup and fix spacing (again)

zoecarver commented:
cc @atrick

atrick commented on lib/SILOptimizer/Utils/Existential.cpp:
```diff
@@ -213,6 +213,15 @@ OpenedArchetypeInfo::OpenedArchetypeInfo(Operand &use) {
+    auto firstUse = *instance->getUses().begin();
+    if (auto *store = dyn_cast<StoreInst>(firstUse->getUser())) {
+      openedVal = store->getSrc();
+    }
```
The existential value stored on the stack here might not be the same value on the stack when it is opened.

Also, the value stored on the stack might no longer be live at the apply site.

There's a whole bunch of complexity in `getAddressOfStackInit` and `canReplaceCopiedValue` to deal with on-stack existentials.

zoecarver replied:
Hmm, I think I understand your first comment but, could you elaborate on the following?
> Also, the value stored on the stack might no longer be live at the apply site.

I've updated line 222 to use `getAddressOfStackInit(..., firstUse->getUser(), ...)` instead of a `dyn_cast<StoreInst>(...)`. Do you think this will be sufficient?

atrick replied:
In the abstract, we want to avoid producing a sequence like
```
r = open_existential_ref
destroy(r)
apply(r)
```
That shouldn't actually be a problem though because we're not in OSSA form so there will only be 'retain/release_value(r)' operations. As long as none of the retains get eliminated during this transformation, it should be safe.

A lot of the complexity in the existing code comes from the fact that it's handling 'open_existential_addr', which *can* be destroyed with 'destroy_addr'.

The main thing here is to prove that the stored value is the value used by the apply. One way to do that is to recognize all uses of the alloc_stack and make sure only one of them is a store. getStackInitInst already does that, but only handles copy_addr, not store.

Please try to write some .sil test cases to for corner cases.

zoecarver replied:
Thank you, that was extremely helpful. 

> A lot of the complexity in the existing code comes from the fact that it's handling 'open_existential_addr', which can be destroyed with 'destroy_addr'.

The constructor I modified does handle `open_existential_addr` but the specific optimization  I'm working on should only handle `open_existential_ref`, right?

> The main thing here is to prove that the stored value is the value used by the apply. One way to do that is to recognize all uses of the alloc_stack and make sure only one of them is a store. 

I've added more checking around the above code but, I suspect it isn't quite right yet. Should the stack alloc instruction ever be used in more than the initial store, the apply, and then the dealloc instructions? If not it may make it a bit easier to filter out what we can optimize. But that might be too harsh of a filter. 

Anyway, I've added many more tests and have uncovered several bugs. Right now I'm in the process of testing those bugs against master to see if they're my issue and if so, fixing them. I'm going to be pretty busy this week with other things so, turn around on this patch might be a bit slow.

zoecarver replied:
@atrick I've added a dominance order check which should ensure that there aren't those kinds of issues. I've also added one .sil test. I plan on writing some more, what other edge cases should I cover?

atrick replied:
If you want to generalize this pattern match, I think it should be part of the existing utilities so there's no new code in this function. `getStackInitInst` should be extended so it can return a store. It already checks for dominance of `ASIUser`. It's a bit more general than your new code. Whether it should set `isCopied` in that case is very subtle. The purpose of that flag is to make sure the original value isn't destroyed before the apply that needs to use it. In the case of a store it's simpler of we do not set `isCopied`, but then we should assert that the store's ownership qualifier is Unqualified (non-OSSA) and  comment when we support OSSA here, we need to insert a new copy of the value before the store (and make sure that copy is destroyed when replacing the apply operand).

FYI: in non-OSSA it should just work. If we have
```
retain val
store val to address
destroy val
use address
destroy_addr
```
It's actually ok in non-OSSA to reuse the original value like this:
```
retain val
store val to address
destroy val
use val
destroy_addr
```

zoecarver replied:
Good idea-- that's much cleaner. Also, thanks for the information about OSSA/non-OSSA and when copies are needed. Very interesting.

atrick commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -472,8 +472,8 @@ SILValue swift::castValueToABICompatibleType(SILBuilder *builder,
-  assert(srcTy.isAddress() == destTy.isAddress()
-         && "Addresses aren't compatible with values");
+  if (srcTy.isAddress() == destTy.isAddress())
+    return nullptr;
```
Is this change intentional?

zoecarver replied:
Yes. This assertion was triggered after the above change. I suspect some other optimization caused the store instruction to trigger this assertion. This was the simplest fix but, maybe I should try to do some more digging.

atrick replied:
It's not part of this APIs contract to return nullptr and it doesn't seem to be anticipated by its callers.

zoecarver replied:
I am now thinking that this assertion shouldn't exist. The next condition `srcTy.isAddress() && destTy.isAddress()` can never succeed if this assertion is triggered (if `bool != bool` then `bool && bool == false`).

atrick replied:
This is valid code
`assert(A == B); if (A && B)`

zoecarver replied:
Yes, you're right. I got confused for a second :P 

I'm not sure what the best solution is, but it seems to be OK when I remove the assertion (not a reason on its own to remove the assertion, though).

zoecarver committed:
Remove assertion all together and use getAddressOfStackInit instead of dyn_cast to StoreInst

zoecarver committed:
Fix based on review and add more tests

zoecarver committed:
Fix tests

zoecarver committed:
Ensure that only correct patterns are optimized

zoecarver committed:
Update comment

zoecarver committed:
Remove redundant protocols

zoecarver commented on test/SILOptimizer/devirtualize_protocol_composition.swift:
```diff
@@ -0,0 +1,138 @@
+  return shouldOptimize1(c)
+}
+
+// TODO: create SR -- this causes a crash on master too
```
[Here's an example.](https://swift.godbolt.org/z/FOPey2) This crashes because of the issue #28012 resolved but even after that commit is applied, it still crashes because of another issue.

atrick replied:
Great catch. Now that you've been looking at this code, can you see what the problem is?

zoecarver replied:
I'll take another look and see if I can figure it out.

zoecarver commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -907,7 +907,7 @@ SILCombiner::buildConcreteOpenedExistentialInfoFromSoleConformingType(
       PD = archetypeTy->getConformsTo()[0];
     } else if (ArgType.isExistentialType() && !ArgType.isAnyObject() &&
-               !SwiftArgType->isAny()) {
+               !SwiftArgType->isAny() && SwiftArgType->getAnyNominal()) {
```
Woops, didn't rebase off master.

zoecarver committed:
Merge branch 'master' into fix/substitution-map-composition

zoecarver committed:
Merge branch 'master' into fix/substitution-map-composition

zoecarver committed:
Check dominance order and add .sil test

zoecarver committed:
Remove .dump in unrelated file

zoecarver committed:
Move relevant part of sil test to the top of the file

zoecarver commented on test/SILOptimizer/devirtualize_protocol_composition_two_applies.sil:
```diff
@@ -0,0 +1,301 @@
+
+public func test(x: B, y: C) -> Int
+
+// CHECK-LABEL: sil shared [noinline] @${{.*}}impl{{.*}}
```
Here's the relevant bit of this test.

zoecarver committed:
Rename test

zoecarver commented:
@atrick friendly ping.

atrick commented:
@zoecarver sorry I keep getting side tracked. This is next on my list to review.

zoecarver commented:
@atrick don't worry about it. Thanks for the update.

atrick commented on test/SILOptimizer/devirtualize_protocol_composition_two_stores.sil:
```diff
@@ -0,0 +1,301 @@
+  dealloc_stack %5 : $*@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  return %8 : $Int
+} // end sil function 'impl'
+
```
We certainly want as many optimizer issues as possible covered by SIL tests, but the tests themselves should be minimal. The maintenance burden of these tests is a primary reason that it's so difficult to make SIL changes. It seems like everything after this point should be deleted, except maybe the SIL declaration of "foo". You don't need to include bodies of called SIL functions.

zoecarver replied:
You're right. I removed much of this sil test. Any other test cases I should cover?

atrick reviewed (changes requested):
It looks like you put a lot of work into this. Thanks. There are a few small cleanups below and an optional suggestion for generalizing the implementation.

zoecarver commented:
@atrick thank you for the review, I really appreciate all the info you gave me here. I might not get to updating this until next week, just FYI.

zoecarver committed:
Merge branch 'master' into fix/substitution-map-composition

zoecarver commented on lib/SILOptimizer/Utils/Existential.cpp:
```diff
@@ -213,6 +213,39 @@ OpenedArchetypeInfo::OpenedArchetypeInfo(Operand &use) {
+    for (auto *use : instance->getUses()) {
+      if (auto *foundStore = dyn_cast<StoreInst>(use->getUser())) {
+        store = foundStore;
+      } else if (auto *foundApply = dyn_cast<ApplyInst>(use->getUser())) {
```
Do you mean `use->getUser() == user`? How would that check be useful? Wouldn't it fail if `use` (the argument) and `use` (the for-loop element) were different uses of the `alloc_stack` instruction?

I'm going to rename some of these variables, I didn't realize there were two `use` variables.

zoecarver committed:
Update getAddressOfStackInit and getStackInitInst to support store instructions

zoecarver committed:
Formatting

zoecarver committed:
Cleanup tests

zoecarver commented:
@atrick no rush but, if you have time, could you re-review this? I think I've addressed all your comments.

atrick commented on lib/SILOptimizer/Utils/Existential.cpp:
```diff
@@ -111,6 +111,19 @@ static SILInstruction *getStackInitInst(SILValue allocStackAddr,
       }
       continue;
     }
```
The comments above `getStackInitInst` should reflect your improvements.

atrick commented on lib/SILOptimizer/Utils/Existential.cpp:
```diff
@@ -195,6 +211,9 @@ static SILValue getAddressOfStackInit(SILValue allocStackAddr,
   if (auto *CAI = dyn_cast<CopyAddrInst>(initI))
     return CAI->getSrc();
 
```
`getAddressOfStackInit` doesn't make sense any more as an API. I think it should just be inlined into `OpenedArchetypeInfo`'s ctor since it seems entirely specific to that context. The the nonsense API name and invalid comments can just be deleted.

zoecarver replied:
Agreed. That's much better. Thanks.

atrick commented on test/SILOptimizer/devirtualize_protocol_composition_two_stores.sil:
```diff
@@ -0,0 +1,114 @@
+  store %3 to %5 : $*@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  dealloc_stack %5 : $*@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  return %8 : $Int
+} // end sil function 'impl'
```
It's generally good habit to also mark the function end with `CHECK-LABEL`. It catches any mistakes in the next CHECK-LABEL and cuts down on confusing `lit` errors.

zoecarver replied:
Good to know. Thanks.

atrick commented on test/SILOptimizer/devirtualize_protocol_composition_two_stores.sil:
```diff
@@ -0,0 +1,114 @@
+} // end sil function 'foo'
+
+// test(x:y:)
+sil @test :$@convention(thin) (@guaranteed B, @guaranteed C) -> Int {
```
I'm unclear as to why the body of `@test` is included but nothing is checked. Is it just to illustrate how `impl` might be called? You could do that in a comment. Or is the SIL actually necessary for the test? Please clarify with a comment, or just remove the SIL body if it doesn't add anything to the test.

zoecarver replied:
Yes, `test` is required because it generates a specialization of `impl`. Without `test`, `impl` isn't optimized (in the way we want) because it doesn't know what parameters it is getting. I'll add some comments to make it more clear what's happening.

atrick reviewed (commented):
Thanks again. Sorry to delay. This looks robust now. I just have a few suggestions about clarity.

zoecarver committed:
Inline/remove getAddressOfStackInit and update tests to make them more clear

zoecarver commented on lib/SILOptimizer/Utils/Existential.cpp:
```diff
@@ -207,11 +205,16 @@ OpenedArchetypeInfo::OpenedArchetypeInfo(Operand &use) {
+    if (auto *initI = getStackInitInst(instance, user, isOpenedValueCopied)) {
+      if (auto *IEA = dyn_cast<InitExistentialAddrInst>(initI))
+        // TODO: this case doesn't need to exist beacuse openedVal will never be used
+        openedVal = IEA;
```
@atrick any reason I shouldn't remove this if-statement?

atrick replied:
I would replace this with a comment that the init_existential_addr case is not handled here because it isn't considered an "opened" archetype, and that case should be handled by ConcreteExistentialInfo instead.

zoecarver commented:
Thanks for the review! 

I found one more test that I need to update (so don't run the bots yet).

zoecarver committed:
Abort if not address type

zoecarver committed:
Support store instructions in createApplyWithConcreteType

Supporting stores in more places means that more code can be _completely_ devirtualized which also means other optimizations (e.g. inlining) can also happen.

Also, updates tests.

zoecarver commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -817,9 +817,15 @@ struct ConcreteArgumentCopy {
-                                 IsInitialization_t::IsInitialization);
-    return ConcreteArgumentCopy(origArg, CAI);
+
+    if (CEI.ConcreteValue->getType().isAddress()) {
```
I originally tried to early-exit here if we didn't have an address but, it turned out to be easier just to generalize this a bit to work with stores as well.

zoecarver commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -882,16 +888,22 @@ SILInstruction *SILCombiner::createApplyWithConcreteType(
+      auto argSub =
+          ConcreteArgumentCopy::generate(CEI, Apply, ArgIdx, BuilderCtx);
+      if (argSub) {
+        UpdatedArgs = true;
```
AFAICT `UpdatedArgs` was never used before. I might clean this up a bit.

zoecarver commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -791,8 +791,8 @@ struct ConcreteArgumentCopy {
 
-    if (!paramInfo.isConsumed())
-      return None;
+//    if (!paramInfo.isConsumed())
```
I don't really see a reason for this to be here. After removing it, I didn't get any errors. I'll run the whole test suit to make sure it's OK. @atrick do you see any reason this should remain?

zoecarver replied:
That's all good to know. I originally thought this was necessary to prevent regression but after I realized the sil test wasn't checking the specialized function this no longer seems necessary. I've added it back.

zoecarver commented:
After updating `OpenedArchetypeInfo` to handle `store` instructions. I also needed to update `createApplyWithConcreteType` and `ConcreteArgumentCopy` to handle stores. So, this patch got a bit bigger (but, I think it would be more hassle to make it two patches).

zoecarver committed:
Add comments, remove commented code, and other general cleanup

zoecarver committed:
Update devirtualize_(...)_two_stores RUN to use sil-opt

zoecarver commented on test/SILOptimizer/existential_specializer_indirect_class.sil:
```diff
@@ -8,74 +8,68 @@ protocol ClassProtocol: AnyObject { func method() }
 
-// CHECK-LABEL: sil shared @$s28test_indirect_class_protocolTf4e_n : $@convention(thin) <τ_0_0 where τ_0_0 : ClassProtocol> (@in τ_0_0) -> ()
-sil hidden @test_indirect_class_protocol : $@convention(thin) (@in ClassProtocol) -> () {
+// CHECK-LABEL: sil [signature_optimized_thunk] [always_inline] @test_indirect_class_protocol : $@convention(thin) (@in ClassProtocol) -> ()
```
cc @jckarter looks like you wrote this test so, I should probably ping you before completely changing it. Does this update seem reasonable?

jckarter replied:
I don't remember this test particularly, but it looks like the existential specializer is no longer applying to this function. Maybe that's expected because your changes are devirtualizing away uses of the existential to begin with? It'd be good to double check.

zoecarver replied:
Thanks for pointing that out. The specializations are still being generated but after they're inlined they are removed. I'm checking the wrong function here. I'll update the `RUN` command to exclude dead-function-elimination and then check the specialized function instead.

atrick commented on lib/SILOptimizer/Utils/Existential.cpp:
```diff
@@ -111,6 +111,19 @@ static SILInstruction *getStackInitInst(SILValue allocStackAddr,
+        SingleWrite = store;
+        // When we support OSSA here, we need to insert a new copy of the value
+        // before `store` (and make sure that the copy is destroyed when
+        // replacing the apply operand).
```
My previous comment might have been misleading because I didn't realize we're inserting a new destroy. In OSSA, the store will do an implicit copy. But, currently, you need to copy the stored value, probably using a retain_value before the store.

To keep the ownership clean what probably ends up happening is this *I think*:

We have an original existential argument value and a new concrete argument value.

The original existential argument value is removed from a "consuming" call operand, and a new destroy_addr of that value is generated in its place (right after the call).

The copy of the new concrete argument value is emitted. That copy is then consumed by the call argument that it is assigned to.

That will be fairly straightforward to update for OSSA later.

zoecarver replied:
Here's a question I've been wanting to ask for a while: what's the relationship between OSSA and reference counting? Is it handled differently in OSSA?

Thanks for laying it all out. That mostly makes sense except for the part about adding a new `destroy_addr` instruction. Currently, the destroy_addr is for the _new_ value (the "tempArg") are you saying it should instead (or in addition) be for the old argument? 

Again, thank you for all the help and information you've given me in this patch. I'm very grateful.

atrick replied:
OSSA is a more strict form of SIL. That's all. It means that each SSA value has its own ownership. So you can't arbitrarily retain/release values and hope they all balance out. A value can only be consumed once (by a destroy_value or owned argument) and cannot be used after it is consumed.

It looks to me like the new destroy at the call site is only created when a `ConcreteArgumentCopy` was generated for the argument. Since ConcreateArgumentCopy always copies/retains, the new destroy will balance that out.

There's another issue I was worried about, which is a totally separate issue from ConcreteArgumentCopy: getStackInitInst will discover the source of the store instruction. We use that as the new ConcreteValue. But, we don't set `isConcreteValueCopied`. So, for stored values, we never check `canReplaceCopiedArg`. The purpose of canReplaceCopiedArg is to ensure that the concrete value is "available" at the call site. So you could have:

```
address = alloc_stack
retain concreteValue // copy to stack
store concreteValue to address
release concreteValue // concreteValue out of scope
apply (@guaranteed address)
use address
destroy address
```

Optimizes to:
```
address = alloc_stack
retain concreteValue // copy to stack
store concreteValue to address
release concreteValue // no longer makes sense
apply (@guaranteed concreteValue)
use address
destroy address
```

That's kind of an abuse of SIL, but it's fine for non-OSSA, and in
fact optimal because ARCCodeMotion will turn it into this:

```
address = alloc_stack
store concreteValue to address
apply (@guaranteed concreteValue)
use address
destroy address
```


But with OSSA it will be:

```
address = alloc_stack
store [init] concreteValue to address
destroy concreteValue
apply (address)
use address
destroy address
```

Optimizes to:
```
address = alloc_stack
copied = copy_value concreteValue
store [init] concreteValue to address
destroy concreteValue
apply (concreteValue)
destroy copied
use address
destroy address
```

I think this PR is fine as-is since you've added a comment for how to handle OSSA later.

zoecarver replied:
Interesting. Really good to know, thank you for taking the time to explain. I'm happy to work on updating more parts of the codebase to use OSSA if you or someone else can help me get started (the knowledge transfer may be more effort than it's worth, though).

atrick commented on test/SILOptimizer/existential_specializer_indirect_class.sil:
```diff
@@ -8,74 +8,68 @@ protocol ClassProtocol: AnyObject { func method() }
-  // CHECK-NEXT: destroy_addr %3
+  // CHECK-NEXT: %1 = load %0
+  // CHECK-NEXT: strong_release %1
+  // CHECK-NEXT: strong_release %1
```
Can you understand why this function isn't specialized anymore?

A single `destroy_addr` is now replaced with two releases? Maybe I'm missing something be that doesn't look right.

zoecarver replied:
It actually is being specialized. You can see that because the inlined version is different. Anyway, if I disable dead-function-elimination the specialized functions come back. I'll do that.

zoecarver replied:
> A single destroy_addr is now replaced with two releases? Maybe I'm missing something be that doesn't look right.

Agreed. It's wrong. The reason for this is in `createApplyWithConcreteType` a new destroy addr instruction is created for each element in `concreteArgCopies`. These are promoted to be `strong_release`s but, a `strong_retain` is never created. If I create a `strong_retain` in `ConcreteArgumentCopy::generate` everything seems to go back to normal but, I don't think those instructions are comparable. Maybe the `destroy_addr` instruction could be replaced with a `strong_release`? Also, how did this work with `copy_addr`?

I really appreciate your patience and explanation. Sorry for all the back and forth we have had (and I'm sure will have). I don't know much about swift's reference counting (I'm eager to learn more, though).

atrick reviewed (changes requested):
More comments/questions...

zoecarver commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -957,8 +968,7 @@ SILInstruction *SILCombiner::createApplyWithConcreteType(
     auto cleanupLoc = RegularLocation::getAutoGeneratedLocation();
     for (ConcreteArgumentCopy &argCopy : llvm::reverse(concreteArgCopies)) {
       cleanupBuilder.createDestroyAddr(cleanupLoc, argCopy.origArg);
```
Yes, something around here needs to be updated. See my below comments. I'm curious how this worked with before this patch. Wouldn't the same issue occur for `copy_addr`?

zoecarver commented:
I've been working through miscompiles after applying this patch and building the standard library. Currently, I'm stuck on some optimization infinite loop, and I'm trying to track down a reproducer. Is there any way to pass `-Xllvm -sil-print-all` (or something similar) to the standard library build?

zoecarver commented:
Alrighty. I've spent a bit more time debugging thing and it seems like there isn't actually an infinite loop but, `GenericSignatureBuilder` is taking an exceedingly long time to `evaluate`/`computeGenericSignature`. Any ideas about why this is? I haven't yet found anything that seems like it would be causing the slowness but, I'll keep looking/debugging.

zoecarver committed:
Stash debugging changes

atrick commented:
> I've been working through miscompiles after applying this patch and building the standard library. Currently, I'm stuck on some optimization infinite loop, and I'm trying to track down a reproducer. Is there any way to pass -Xllvm -sil-print-all (or something similar) to the standard library build?

You can run `sil-opt Swift.swiftmodule` to see the final SIL for each function after optimization. Or you can grab the `swift -frontend` command line that builds the stdlib and trace a particular function prefix using `-Xllvm '-sil-print-only-functions=xxx'`, which should also print specialization of the function. Or just instrument the compiler yourself using `if F->hasName('xxx') llvm::dbgs() << xxx`

zoecarver commented:
Thanks for all those suggestions! I didn't realize you could give `sil-opt` the whole module. That might be just what I need to narrow this down.

zoecarver committed:
Remove debug code and update check for successful conversion of
arguments.

zoecarver committed:
Re-add support for try_apply instruction and update tests.

Re-work tests based on updated optimizations and check specialized
functions.

zoecarver commented:
The plot thickens: now everything (seems to) work in debug mode _but_, in release mode (such as the buildbot preset) there's an infinite loop (or _really_ slow operation) introduced.

zoecarver committed:
Stash optimization debugging

zoecarver committed:
Update how createApplyWithConcreteType checks if it can update Apply
args

createApplyWithConcreteType used to update Apply unconditionally. There
may have been cases prior to supporting store instructions that
miscompiled because of this but, there are certainly cases after
supporting store instructions. Therefore, it is important that the apply is updated only when the substitution can accept the new args.

zoecarver committed:
Format

zoecarver committed:
Add a `madeUpdate` check to `createApplyWithConcreteType` to prevent
infinite loop

zoecarver commented:
I'll run the tests on my machine tonight but (fingers crossed after a whole load of debugging this week) I think everything is _finally_ working properly.

zoecarver committed:
Manually check if types are the same when checking for change and update
tests

zoecarver commented:
Alright. All `SILOptimizer` tests are passing. @atrick can you re-review/trigger the bots?

zoecarver committed:
Format

atrick commented:
@swift-ci test

atrick commented:
@swift-ci benchmark

atrick commented:
@swift-ci test source compatibility

zoecarver commented:
It looks like this optimization didn't improve any benchmarks 😞 It improved silgen size but not assembly (even with the added tests). Maybe the llvm optimizer was already doing this optimization so, it might still be beneficial in more complex cases. Is there any way to see the performance deltas (code size, probably) from the stdlib? 

Sorry this wasn't a more beneficial optimization, I still appreciate all your help and time working on it with me.

atrick commented:
@zoecarver The benchmark that improves should actually be included in this PR. I don't remember the inception of this bug. I think someone was actually using protocol composition (pretty common in practice I think) and ran into a correctness issue. I think we fixed that issue by bailing out and deferred optimizing it correctly by filing this bug. It looks like I attached the .swift test case to this bug. It shouldn't be too hard to create a benchmark out of that test case.

zoecarver committed:
Add benchmark for devirtualization performance measurements

zoecarver committed:
Update added tests to require objec interop

zoecarver commented:
@atrick OK, great. I've added a benchmark (based on the example) and it seems to improve drastically after this patch is applied. Also, the linux failures should be fixed (required objc). 

Could you re-trigger the bots? Thanks again :)

atrick commented:
@swift-ci test

atrick commented:
@swift-ci benchmark

zoecarver committed:
Use blackHole to make sure benchmark isn't optimized away

zoecarver commented:
The build failure says it is because of swiftpm. I don't think that's because of this patch (I could be wrong, though). @atrick what should I do to fix the swiftpm error, or should I not worry about it? Also, I've updated the benchmarks to use `blackHole`. I can verify that _both_ before, and after this patch, the benchmarks take time to run.

atrick commented:
@zoecarver unfortunately not everyone runs regular PR testing after making functional changes! So the default answer is keep trying until the tests pass.

atrick commented:
@swift-ci test

zoecarver commented:
OK, at least I don't have to fix anything :P

zoecarver commented:
Looks like all the tests passed. Maybe you could trigger the benchmarks again and if it looks good to you, can you merge it for me? 

Thanks again for all the help with this patch, @atrick!

atrick commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -914,7 +924,29 @@ SILInstruction *SILCombiner::createApplyWithConcreteType(
   }
 
-  if (!UpdatedArgs) {
+  bool canUpdateArgs, madeUpdate;
```
This whole section of new code was confusing at first. I *think* what it's doing is determining both whether it is possible to rewrite the call at all and whether at least one argument has changed. If not, avoid creating a new call and delete any temporary instructions to avoid the top-level SILCombine driver from infinitely attempting to optimize. If that's right, then the code looks good but please explain that in a comment before new code.

I'm not sure why it's a written as a closure. The results need to be exposed anyway and there's not much else to hide here.

zoecarver replied:
I'll factor it out and add a comment. Thanks.

atrick reviewed (approved):
I requested a comment, but otherwise this PR looks good now.

zoecarver committed:
Merge branch 'master' into fix/substitution-map-composition

zoecarver committed:
Break argument checking out of a lambda and add comments

zoecarver commented:
@atrick thanks, as I've said before, I really appreciate all your reviews of this patch. I updated it based on your comment.

atrick commented:
@zoecarver migrating passes to OSSA is one of the most useful things that can be done. I'm not sure which pass is easiest to start with. It's something we'll be figuring over the next few months. @gottesmm is looking into it.

atrick commented:
@swift-ci smoke test

atrick commented:
@zoecarver thanks for iterating on this to get it right!

zoecarver commented:
> @zoecarver thanks for iterating on this to get it right!

Of course. 

> @zoecarver migrating passes to OSSA is one of the most useful things that can be done. I'm not sure which pass is easiest to start with. It's something we'll be figuring over the next few months. @gottesmm is looking into it.

I think I remember @gottesmm saying that it would be hard to find a good pass for me to work on but, let me know if that changes, or if there are other things that I can do that would be helpful.

zoecarver commented:
@atrick could you also run the benchmarks? The benchmark I added should now work properly.

atrick commented:
@swift-ci benchmark

zoecarver commented on benchmark/single-source/DevirtualizeProtocolComposition.swift:
```diff
@@ -0,0 +1,65 @@
+  return sum
+}
+
+func shouldOptimize1<T>(_ x: ClassA<T> & ProtocolA, count: Int) -> Int {
```
I'll fix the names here.

zoecarver commented on benchmark/single-source/DevirtualizeProtocolComposition.swift:
```diff
@@ -0,0 +1,65 @@
+
+@inline(never)
+public func entryPoint1(c: ClassB) -> Int {
+  return shouldOptimize1(c, count: 25)
```
The idea here is that before this patch `shouldOptimize1` counts to 2500 and after this patch, the number 2500 is inlined directly into `entryPoint1` as an `integer_literal`. I may need to add more work to the unoptimized version. 

When I ran the benchmarks on my computer, I thought they were significantly different but, now I'm thinking I may have misread them. The results from running the benchmarks locally:
**Before**:
```
215,DevirtualizeProtocolComposition,1,34914,34914,34914,0,34914
```
**After**:
```
215,DevirtualizeProtocolComposition,100,35,46,36,1,36
```
Anything else I could do to improve this benchmark?

atrick replied:
Looks like it's too well optimized to measure reliably now. You'll need enough iterations in an "outer" loop such that the measured time is significantly larger than the setup time.

atrick replied:
Oh, another thing. Please make the non-optimized benchmark faster. We don't want to slow down -Onone benchmarking and it's more than sufficient for one benchmark to slow down by 2x or 10x if this optimization breaks. 10,000x is a bit outrageous.

atrick replied:
It might help to look at how some of the other devirtualization benchmarks are setup. If you don't know which ones, you could disable the pass and find out. It's nice if the benchmark mimics a realistic scenario, even if it is just a microbenchmark

zoecarver committed:
Update benchmark to be more similar to other devirtualization benchmarks

atrick commented:
@swift-ci smoke test

atrick commented:
@swift-ci benchmark

futurejones commented:
Can one of the admins verify this patch?

zoecarver commented:
@atrick friendly ping. Anything else I need to do with this?

atrick commented:
@swift-ci smoke test

zoecarver commented:
Looks like a swiftpm issue.

theblixguy commented:
@swift-ci please smoke test Linux

zoecarver commented:
@theblixguy thanks!

atrick commented:
@swift-ci smoke test linux
